src/lib/checkpoint/cp_mgr.cpp

```diff
@@ -205,6 +214,54 @@ void CPManager::cleanup_cp(CP* cp) {
     }
 }
 
+//////////////////////////////////////// CP Guard class ////////////////////////////////////////////
+CPGuard::CPGuard() {
+    if (t_cp_stack.empty()) {
+        // First CP in this thread stack.
+        m_cp = hs()->cp_mgr().cp_io_enter();
+    } else {
+        // Nested CP sections
+        m_cp = t_cp_stack.top();
+        hs()->cp_mgr().cp_ref(m_cp);
+    }
+    t_cp_stack.push(m_cp);
```

Comment:
for nested cp case, here we are pushing same cp to stack, right? There is no problem with that, just trying to make sure I understand correctly.

Reply:
yes, either we need to push or maintain same pointer and count pair. This seems less penalty and also code looks much cleaner this way (I had a version like you mentioned - but it seems not clean). Also this case is generally rare to be more than 2 in this stack.